### ISSUE TITLE
Running 'Hashicorp Terraform' Formatter hangs

### ISSUE THREAD

**Issue by kylehutchens** [hashicorp/vscode-terraform#878]
## Versions

This bug is reproducible in:
 - [ ] the **latest** version of the extension (below)
 - [ ] the **latest** version of the language server (below)

### Extension

```
v2.17.0
```

### Language Server

```
0.25.0
```

### VS Code

```
1.63.2
```

### Operating System

```
macOS 11.5.2
```

### Terraform Version

```
1.0.1
```

### Steps To Reproduce

When attempting to save tf I get a pop up in vscode "Running 'Hashicorp Terraform' formatter" that will just hang and not save my file until I cancel the task. Nothing will output to the hashicorp terraform output view. Restarting vscode will usually get it working again temporarily. New issue to me but I have tried so many things to resolve and nothing has worked.

### Expected Behavior

The terraform formats
### Actual Behavior

 It hangs
### Additional context

I don't see a terraform-ls output view but if you can direct me where to find it I will get you logs from the language server. I saw that there was another issue opened about this but no movement has happened on it in months and I fear no one is paying attention to that issue anymore. [#650](https://github.com/hashicorp/vscode-terraform/issues/650)

**Comment by jpogran** [comment]
Hi @kylehutchens, thank you for the report and the provided environment information. Can you please also provide the following?

```
editor.defaultFormatter
editor.formatOnSave
editor.formatOnType
editor.formatOnPaste
editor.formatOnSaveMode
editor.codeActionsOnSave
```

If you have specified any settings specifically for the Terraform language (by setting "[terraform]":{}), please include those as well.

To see the terraform-ls log, you can open the Output Pane and select the `HashiCorp Terraform` view as show below:

![image](https://user-images.githubusercontent.com/272569/147266957-75bf2d06-dff0-46f0-a13b-e983f49f2458.png)

I did update #650 with additional information that may be helpful for you.

**Comment by kylehutchens** [comment]
> Hi @kylehutchens, thank you for the report and the provided environment information. Can you please also provide the following?
> 
> ```
> editor.defaultFormatter
> editor.formatOnSave
> editor.formatOnType
> editor.formatOnPaste
> editor.formatOnSaveMode
> editor.codeActionsOnSave
> ```
> 
> If you have specified any settings specifically for the Terraform language (by setting "[terraform]":{}), please include those as well.

I assume you're asking what I have in my settings.json for that? I will include the relevant values here, any omitted are not present and left as default.
```
  "[terraform]": {
    "editor.formatOnSave": true,
  },
  "[terraform-vars]": {
    "editor.formatOnSave": true,
  },
  ```

> To see the terraform-ls log, you can open the Output Pane and select the `HashiCorp Terraform` view as show below:
> 
> ![image](https://user-images.githubusercontent.com/272569/147266957-75bf2d06-dff0-46f0-a13b-e983f49f2458.png)
> 
> I did update #650 with additional information that may be helpful for you.

Got it, that's where I grabbed the terraform-ls version number from but I have seen other screenshots showing a terraform-ls output view while searching for the solution to this issue. When the formatting stops working and thus the hang on running formatter I get no output to the pane so I am guessing the language server may be crashing or not responding?

Thank you for responding on the other issue as well I wasn't sure if that was something I should create a new issue for or not but it seemed like enough time and versions had passed that that issue was not totally relevant anymore.

**Comment by jpogran** [comment]
In older versions of the extension that view was called `terraform-ls`. We recently renamed it to `HashiCorp Terraform` for easier identification. I will update the `New Issue` template with the new wording.

Yes, if there is no additional text displayed in that view when formatting happens, then the language server is still processing and hasn't returned, which means it may be hung or non-responsive. 

Creating a new issue is fine, don't worry about it. If it was a duplicate, we can always link and close. In this case, it's helpful to keep these reports separate, as sometimes it's situation specific and not general.

**Comment by kylehutchens** [comment]
I looked through your post on [#542](https://github.com/hashicorp/vscode-terraform/issues/542) and changed my `settings.json` to this[1] though I didn't change anything with the formatting. Is there any where else I can grab logs or information from that would be helpful to you? 

[1]
```
  "[json]": {
    "editor.defaultFormatter": "vscode.json-language-features"
  },
  "[terraform]": {
    "editor.defaultFormatter": "hashicorp.terraform",
    "editor.formatOnSave": true,
    "editor.formatOnSaveMode": "file" 
  },
  "[terraform-vars]": {
    "editor.defaultFormatter": "hashicorp.terraform",
    "editor.formatOnSave": true,
    "editor.formatOnSaveMode": "file"
  },
  ```

**Comment by jpogran** [comment]
The only other thing at the moment is to describe the project you're opening. Single folder with a few modules? Large project with many modules? Hundreds of modules? Will be able to dig in more after the holiday

**Comment by kylehutchens** [comment]
We use a monolithic architecture repo approach and are easily in the hundreds of modules inside of the repository but broken out into many smaller micro-services so no one tf directory is more than 10 modules or so.

**Comment by kylehutchens** [comment]
Hey @jpogran, hope your holidays went well. Wondering if you had some time to circle back to looking at this with me. Thanks!

**Comment by jpogran** [comment]
No immediate recommendations at the moment.

We have a new release of the terraform-ls (0.25.1) binary going out today with an interim fix for cpu usage spikes that may be related to your issue. It's targeted at large solutions where the LS tries to parse module information and uses up alot of CPU. This may be preventing the formatting call from being picked up in the queue.

Once we get that release out I can circle back here and dig in further. Please let me know if that release changes any behavior for you.

**Comment by radeksimko** [comment]
Based on some more data we received and reviewed there's a reason to believe that this is triggered by the amount of data being processed by the refcount code lens, which is in turn enabled by default. We plan to [disable it by default](https://github.com/hashicorp/vscode-terraform/pull/893).

Can you try to disable it using the following settings?

```json
"terraform.codelens.referenceCount": false,
```

**Comment by kylehutchens** [comment]
@radeksimko I installed the newest language server, `0.25.1`, and updated my settings to be[1] but I am still having the issue of the formatter hanging on save. Let me know if there is anything else that I can try.

[1]  
```"terraform.codelens.referenceCount": false,
  "terraform.languageServer": {
    "args": [
      "serve"
    ],
  },

  "[terraform]": {
    "editor.defaultFormatter": "hashicorp.terraform",
    "editor.formatOnSave": true,
    "editor.formatOnSaveMode": "file" 
  },
  "[terraform-vars]": {
    "editor.defaultFormatter": "hashicorp.terraform",
    "editor.formatOnSave": true,
    "editor.formatOnSaveMode": "file"
  },
```

**Comment by radeksimko** [comment]
@kylehutchens No, the formatter would never format more than a single file. If it's hanging, it's more likely a symptom of a different problem.

Do you mind sharing the log from the `Output` pane, so we can see what's going on when the formatting is attempted? Ideally the whole log from top to bottom.
![Screenshot 2022-01-06 at 16 28 44](https://user-images.githubusercontent.com/287584/148416041-e85d366d-65e0-4b1b-a5a3-8cb91da6ccb4.png)

You can also [GPG-encrypt](https://keybase.io/radeksimko) it if you're worried about sensitive data in it.

If you're on Unix-based OS (Linux or macOS) - can you collect a CPU profile and share it with us? https://github.com/hashicorp/terraform-ls/blob/main/docs/TROUBLESHOOTING.md#cpu-profiling

You may also send it privately via email to `radek<at>hashicorp.com` if you wish.

**Comment by kylehutchens** [comment]
@radeksimko I sent you the output of the extension but I noticed that nothing made it to the output pane after initial startup. Also when I updated my settings file[1] to make a cpu profile as requested I got this output[2]. I feel like this is not where I would make this change for vscode? 

[1]```
  "terraform.languageServer": {
    "args": [
      "serve -cpuprofile=/tmp/terraform-ls-cpu.prof"
    ],
  },
```
[2]```
Launching language server: /Users/kylehutchens/Library/Application Support/Code/User/globalStorage/hashicorp.terraform/bin/terraform-ls serve -cpuprofile=/tmp/terraform-ls-cpu.prof
[Info  - 2:53:37 PM] Connection to server got closed. Server will restart.
[Info  - 2:53:37 PM] Connection to server got closed. Server will restart.
[Info  - 2:53:37 PM] Connection to server got closed. Server will restart.
[Info  - 2:53:37 PM] Connection to server got closed. Server will restart.
[Error - 2:53:37 PM] Connection to server got closed. Server will not be restarted.
```

**Comment by radeksimko** [comment]
@kylehutchens Each argument in the settings needs to be separated, i.e.

```json
  "terraform.languageServer": {
    "args": [
      "serve",
      "-cpuprofile=/tmp/terraform-ls-cpu.prof"
    ],
  },
```

**Comment by kylehutchens** [comment]
@radeksimko Ah, thank you, I updated the settings file and the language server started and a file was generated at the path but it appears to be empty(opening in text edit is a blank file and finder reports it as being `zero bytes`. I can send it along to you but I don't think it's what you're looking for.

**Comment by jpogran** [comment]
We recently released v2.180 of the extension, which changes `terraform.codelens.referenceCount` to be disabled by default. In certain scenarios the overhead of pre-calculating the number of references for each block in a document causes high CPU utilization. This may be related to your issue, but might not be. I'll be looking into this in more detail later this week.

**Comment by kylehutchens** [comment]
I am still having the issue with extension version `2.18.0`

**Comment by radeksimko** [comment]
@kylehutchens I looked at your log which you sent privately (Thank you!) - and responded with some more questions. I'm not fully convinced you're experiencing the same bug we fixed in 2.18.0 - it may be something completely different, but we won't know without more details I'm afraid.

**Comment by kylehutchens** [comment]
@radeksimko Hey I followed up with you in an email. 

For anyone following along I just explained that my output from a fresh open of vscode takes a long time to start filling up with data but will get so much data in it that I can no longer see the beginning of the output. I will be testing opening a smaller project and see if the size of this repository is giving the extension issues.

**Comment by kylehutchens** [comment]
Follow-up email sent to Radek
```
I went into the repository for one of our modules and I am convinced now that the quantity of modules in the main repo is the culprit. Though even this module repo had some weirdness. I loaded up the repository and edited the spacing to trigger a format when I saved and then saved and the file formatted immediately! But then I tried it again and it hung, I checked the output pane and a few seconds later a large amount of output went through and I attempted to format on save again and it worked that time and the next time I tried it. Please see attached for this output's logs.
```

**Comment by radeksimko** [comment]
We have just released language server `0.25.2` which patches a number of bugs, mainly reducing overhead in various scenarios, such as publishing diagnostics, refreshing code lens or telemetry - which could previously block or delay formatting requests.

With that in mind, I believe we have done all we could based on the data and knowledge available at this point, so I'm going to close this issue.

If anyone still experiences these symptoms (or similar) after upgrading to the latest 2.18.0+ of the extension and 0.25.2 of LS, I would encourage you to file a new issue and attach details, such as log file, CPU profile and situation in which this occurs.

Language server gets updated automatically every 24hours or upon the next VS Code launch/restart, you can find the LS version near the top of the log under `Output -> HashiCorp Terraform`.

Thanks @kylehutchens for all your help.
